src/alignment/sparse.rs

```diff
@@ -0,0 +1,452 @@
+        events.push((x+k,y+k,idx as u32));
+
+        n = max(n, x+k);
+        n = max(n, y+k);
+    }
+    events.sort(); 
+
+
+    let mut max_col_dp: MaxBitTree<PrevPtr> = MaxBitTree::new(n as usize);
+    let mut dp: Vec<(u32, i32)> = Vec::with_capacity(events.len());
+    let mut best_dp = (0,-1);
+
+    for _ in 0 .. events.len() {
+        dp.push((0,0));
+    }
+
+    for ev in events {
+        let p = (ev.2 % matches.len() as u32) as usize;
+        let j = ev.1;
+        let is_start = ev.2 >= (matches.len() as u32);
+        
+
+        if is_start {
+            // Default case -- chain starts at this node
+            dp[p] = (k * match_score, -1);
+
+            // Find best previous chain, and extend. 
+            let best_prev = max_col_dp.get(j as usize);
+            if best_prev.score > 0 {
+                
+                let prev_d = best_prev.d;
+                let cur_d = ev.0 + ev.1;
+                let gap = _gap_open + (cur_d - prev_d) * _gap_extend;
+
+                let reward = k * match_score;
+                let new_score = (best_prev.score + reward).saturating_sub(gap);
+
+                dp[p] = max(dp[p], (new_score, best_prev.id as i32));
+                best_dp = max(best_dp, (dp[p].0, p as i32));
+            }
+        } else {
+            // See if this kmer continues a diffent kmer
+            if ev.0 >= k+1 && ev.1 >= k+1 {
+                match matches.binary_search(&(ev.0-k-1, ev.1-k-1)) {
+                    Ok(cont_idx) => {
+                        let prev_score = dp[cont_idx].0;
+                        let candidate = (prev_score + match_score, cont_idx as i32);
+                        dp[p] = max(dp[p], candidate);
+                        best_dp = max(best_dp, (dp[p].0, p as i32));
+                    },
+                    _ => (),
+                }
+            }
+
+            let prev_frag = PrevPtr::new(dp[p].0, ev.0, ev.1, p, _gap_extend);
+            max_col_dp.set(ev.1 as usize, prev_frag);
+        }
+    }
+
+    let mut traceback = Vec::new();
+    let (best_score, mut prev_match) = best_dp;
+    while prev_match >= 0 {
+        traceback.push(prev_match as usize);
+        prev_match = dp[prev_match as usize].1;
+    }
+    traceback.reverse();
+    SparseAlignmentResult { path: traceback, score: best_score, dp_vector: dp }
+}
+
+
+/// Find all matches of length k between two strings, using a q-gram
+/// index. For very long reference strings, it may be more efficient to use and
+/// FMD index to generate the matches. Note that this method is mainly for 
+/// demonstration & testing purposes.  For aligning many query sequences
+/// against the same reference, you should reuse the QGramIndex of the reference.
+pub fn find_kmer_matches<T: AsRef<[u8]>>(seq1: &T, seq2: &T, k: usize) -> Vec<(u32, u32)> {
+
+    let slc1 = seq1.as_ref();
+    let slc2 = seq2.as_ref();
+
+    let mut set: HashMap<&[u8], Vec<u32>> = HashMap::new();
```

Comment:
Ok, not using qgramindex is totally fine for one-time use. However, I wonder if it would still be beneficial to iterate over and store bit-encoded qgrams instead of byte slices. But this can also be checked later if you like. Also, can you rename everything from kmer to qgram? Just to be consistent with the rest of the library.